File: Lab5/main.py
import numpy as np
import matplotlib.pyplot as plt


def exact_solution(x):
    return np.tan(x) - x


def f(x, u):
    return (u + x) ** 2


def newton_method(x, u_prev, h, tol=1e-8, max_iter=100):
    u = u_prev
    for _ in range(max_iter):
        fu = u - u_prev - h / 2 * (f(x, u_prev) + f(x + h, u))
        dfu = 1 - h / 2 * 2 * (u + x + h)
        delta = fu / dfu
        u -= delta
        if abs(delta) < tol:
            break
    return u


def implicit_trapezoidal(a, b, h, u0):
    n = int((b - a) / h)
    x_values = np.linspace(a, b, n + 1)
    u_values = np.zeros(n + 1)
    u_values[0] = u0
    for i in range(n):
        x = x_values[i]
        u_values[i + 1] = newton_method(x, u_values[i], h)
    return x_values, u_values


def runge_kutta_2(a, b, h, u0):
    n = int((b - a) / h)
    x_values = np.linspace(a, b, n + 1)
    u_values = np.zeros(n + 1)
    u_values[0] = u0
    for i in range(n):
        x = x_values[i]
        k1 = f(x, u_values[i])
        k2 = f(x + h / 2, u_values[i] + h / 2 * k1)
        u_values[i + 1] = u_values[i] + h * k2
    return x_values, u_values


def compute_max_error(x_values, y_values):
    exact = exact_solution(x_values)
    errors = np.abs(exact - y_values)
    return np.max(errors)


if __name__ == "__main__":
    a, b = 0.0, 1.0
    h1, h2 = 0.1, 0.05
    u0 = 0.0

    x_trap_h1, u_trap_h1 = implicit_trapezoidal(a, b, h1, u0)
    x_trap_h2, u_trap_h2 = implicit_trapezoidal(a, b, h2, u0)

    x_rk_h1, u_rk_h1 = runge_kutta_2(a, b, h1, u0)
    x_rk_h2, u_rk_h2 = runge_kutta_2(a, b, h2, u0)

    error_trap_h1 = compute_max_error(x_trap_h1, u_trap_h1)
    error_trap_h2 = compute_max_error(x_trap_h2, u_trap_h2)
    error_rk_h1 = compute_max_error(x_rk_h1, u_rk_h1)
    error_rk_h2 = compute_max_error(x_rk_h2, u_rk_h2)

    p = 2
    R_trap = abs(error_trap_h1 - error_trap_h2) / (2 ** p - 1)
    R_rk = abs(error_rk_h1 - error_rk_h2) / (2 ** p - 1)

    print("Результаты для неявного метода трапеций:")
    print(f"Шаг h1 = {h1}, Максимальная ошибка = {error_trap_h1}")
    print(f"Шаг h2 = {h2}, Максимальная ошибка = {error_trap_h2}")
    print(f"Оценка по правилу Рунге: {R_trap}")
    print()

    print("Результаты для метода Рунге-Кутты (порядок 2):")
    print(f"Шаг h1 = {h1}, Максимальная ошибка = {error_rk_h1}")
    print(f"Шаг h2 = {h2}, Максимальная ошибка = {error_rk_h2}")
    print(f"Оценка по правилу Рунге: {R_rk}")

    # Построение графиков
    plt.figure(figsize=(12, 6))

    # График для шага h1
    plt.subplot(1, 2, 1)
    plt.plot(x_trap_h1, u_trap_h1, label="Неявный метод трапеций (h1)", marker='o', linestyle='--')
    plt.plot(x_rk_h1, u_rk_h1, label="Метод Рунге-Кутты (h1)", marker='x', linestyle='-.')
    plt.plot(x_trap_h1, exact_solution(x_trap_h1), label="Точное решение", linestyle='-', color='black')
    plt.title("Сравнение решений (шаг h1)")
    plt.xlabel("x")
    plt.ylabel("u(x)")
    plt.legend()
    plt.grid()

    # График для шага h2
    plt.subplot(1, 2, 2)
    plt.plot(x_trap_h2, u_trap_h2, label="Неявный метод трапеций (h2)", marker='o', linestyle='--')
    plt.plot(x_rk_h2, u_rk_h2, label="Метод Рунге-Кутты (h2)", marker='x', linestyle='-.')
    plt.plot(x_trap_h2, exact_solution(x_trap_h2), label="Точное решение", linestyle='-', color='black')
    plt.title("Сравнение решений (шаг h2)")
    plt.xlabel("x")
    plt.ylabel("u(x)")
    plt.legend()
    plt.grid()

    plt.tight_layout()
    plt.show()
    

File: Lab5/main.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <iomanip>

double exactSolution(double x) {
    return std::tan(x) - x;
}

double f(double x, double u) {
    return std::pow(u + x, 2);
}

double newtonMethod(double x, double u_prev, double h) {
    double u = u_prev; 
    const double tol = 1e-8; 
    const int max_iter = 100; 
    for (int iter = 0; iter < max_iter; ++iter) {
        double fu = u - u_prev - h / 2 * (f(x, u_prev) + f(x + h, u));
        double dfu = 1 - h / 2 * 2 * (u + x + h); 
        double delta = fu / dfu;
        u -= delta;
        if (std::abs(delta) < tol) break;
    }
    return u;
}

std::vector<double> implicitTrapezoidal(double a, double b, double h, double u0) {   // Неявный метод трапеций
    int n = static_cast<int>((b - a) / h);
    std::vector<double> u(n + 1);
    u[0] = u0;
    for (int i = 0; i < n; ++i) {
        double x = a + i * h;
        u[i + 1] = newtonMethod(x, u[i], h);
    }
    return u;
}

std::vector<double> rungeKutta2(double a, double b, double h, double u0) {   // Метод Рунге-Кутты второго порядка
    int n = static_cast<int>((b - a) / h);
    std::vector<double> u(n + 1);
    u[0] = u0;
    for (int i = 0; i < n; ++i) {
        double x = a + i * h;
        double k1 = f(x, u[i]);
        double k2 = f(x + h / 2, u[i] + h / 2 * k1);
        u[i + 1] = u[i] + h * k2;
    }
    return u;
}

double computeMaxError(const std::vector<double>& x_values, const std::vector<double>& y_values) {
    double max_error = 0.0;
    for (size_t i = 0; i < x_values.size(); ++i) {
        double error = std::abs(exactSolution(x_values[i]) - y_values[i]);
        if (error > max_error) {
            max_error = error;
        }
    }
    return max_error;
}

int main() {
    setlocale(LC_ALL, "ru");
    double a = 0.0, b = 1.0;
    double h1 = 0.1;         
    double h2 = h1 / 2;     
    double u0 = 0.0;       

    std::vector<double> u_trap_h1 = implicitTrapezoidal(a, b, h1, u0);
    std::vector<double> u_trap_h2 = implicitTrapezoidal(a, b, h2, u0);

    std::vector<double> u_rk_h1 = rungeKutta2(a, b, h1, u0);
    std::vector<double> u_rk_h2 = rungeKutta2(a, b, h2, u0);

    std::vector<double> x_values_h1(static_cast<size_t>((b - a) / h1) + 1);
    std::vector<double> x_values_h2(static_cast<size_t>((b - a) / h2) + 1);
    for (size_t i = 0; i < x_values_h1.size(); ++i) {
        x_values_h1[i] = a + i * h1;
    }
    for (size_t i = 0; i < x_values_h2.size(); ++i) {
        x_values_h2[i] = a + i * h2;
    }

    double error_trap_h1 = computeMaxError(x_values_h1, u_trap_h1);
    double error_trap_h2 = computeMaxError(x_values_h2, u_trap_h2);
    double error_rk_h1 = computeMaxError(x_values_h1, u_rk_h1);
    double error_rk_h2 = computeMaxError(x_values_h2, u_rk_h2);

    // Правило Рунге
    double p = 2; 
    double R_trap = std::abs(error_trap_h1 - error_trap_h2) / (std::pow(2, p) - 1);
    double R_rk = std::abs(error_rk_h1 - error_rk_h2) / (std::pow(2, p) - 1);

    std::cout << "Результаты для неявного метода трапеций:" << std::endl;
    std::cout << "Шаг h1 = " << h1 << ", Максимальная ошибка = " << error_trap_h1 << std::endl;
    std::cout << "Шаг h2 = " << h2 << ", Максимальная ошибка = " << error_trap_h2 << std::endl;
    std::cout << "Оценка по правилу Рунге: " << R_trap << std::endl;
    std::cout << std::endl;

    std::cout << "Результаты для метода Рунге-Кутты (порядок 2):" << std::endl;
    std::cout << "Шаг h1 = " << h1 << ", Максимальная ошибка = " << error_rk_h1 << std::endl;
    std::cout << "Шаг h2 = " << h2 << ", Максимальная ошибка = " << error_rk_h2 << std::endl;
    std::cout << "Оценка по правилу Рунге: " << R_rk << std::endl;

    return 0;
}
